### PR TITLE
[sources/path] Add gitignore-like pattern matching and warn on mismatches

### PR DESCRIPTION
**Pull request by behnam**
Add gitignore-like pattern matching logic to `list_files()` and throw
warnings for paths getting different inclusion/exclusion results from
the old and the new methods.

Migration Tracking: <#4268>

### CODE REVIEW

**Comment by behnam** [comment]
This is a WIP, @alexcrichton.

Writing extensive unit test for the `exclude` option, I noticed that there are surprising changes for some patterns. I'm still investigating them, but I'm guessing they could be *expected* gitignore-style behavior, or a bug in `ignore` crate. (If expected, it's going to be annoying for the migration.)

So, the details for the pattern matching needs some work. Would appreciate feedback on overall direction of the patch.

**Comment by behnam** [comment]
Turns out `ripgrep`'s `ignore` also has some issues in edge cases. I have added integration tests that reveal the bugs: https://github.com/BurntSushi/ripgrep/pull/551

I'm going to expand the test and cover more cases and make sure `ignore` package is in good shape before we move forward here.

File: src/cargo/sources/path.rs [hunk 1]
```diff
@@ -85,42 +87,161 @@ impl<'cfg> PathSource<'cfg> {
...
+
+        let glob_include = pkg.manifest()
+            .include()
+            .iter()
+            .map(|p| glob_parse(p))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let glob_should_package = |relative_path: &Path| -> bool {
+            // include and exclude options are mutually exclusive.
+            if no_include_option {
```
Comment (alexcrichton) [inline src/cargo/sources/path.rs]: I think this condition is different from before, right? The `exclude` array is now ignored if `include` is provided? (looks accidental?)

File: src/cargo/sources/path.rs [hunk 2]
```diff
@@ -85,42 +87,161 @@ impl<'cfg> PathSource<'cfg> {
...
+                    pattern.matches_path(relative_path)
+                })
+            }
+        };
+
+        // ignore-like matching rules
+
+        let mut exclude_builder = GitignoreBuilder::new(root);
+        for rule in pkg.manifest().exclude() {
+            exclude_builder.add_line(None, rule)?;
```
Comment (alexcrichton) [inline src/cargo/sources/path.rs]: Just to double check, how can this fail? This may be an error we wish to handle if valid `glob`-crate patterns are invalid gitignore patterns right now

File: src/cargo/sources/path.rs (hunk 1 above)
Comment (behnam) [inline src/cargo/sources/path.rs]: No, it's been in the code for a couple of years:
https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/path.rs#L112-L113

And the docs also clearly say this: http://doc.crates.io/manifest.html#the-exclude-and-include-fields-optional

> The options are mutually exclusive: setting `include` will override an `exclude`.

That's why I suggested (in https://github.com/rust-lang/cargo/issues/4268) to stop that after we're done with the other fixes, and have include work as the first step, and exclude as a second step.

File: src/cargo/sources/path.rs (hunk 2 above)
Comment (behnam) [inline src/cargo/sources/path.rs]: I have not looks at the details, but at the moment, I guess it's bad usage of special characters (which are `*`, `?`, `!`), which, in most cases, result also in bad glob patterns. (Well, basically, every line of gitignore is a glob, with the addition of negation by leading `!`, which cannot fail.)

So, I don't think falling back to a Glob object helps much.

But, maybe we should just throw warnings here and move on. But, I think a bad include/exclude rule is bad enough to break the packaging step.

Also, I don't expect any error to be introduced with the migration. Still need to check if patterns with `!` can throw errors. Let me take a look at that and get back here.

Comment (behnam) [inline src/cargo/sources/path.rs]: Okay, I checked and the only type of error returned here is from `Glob` (`globset::Glob::Error`), which happens in cases like `***`. So, there's no reason to fallback here.

File: src/cargo/sources/path.rs (hunk 1 above)
Comment (alexcrichton) [inline src/cargo/sources/path.rs]: Ah yes, my mistake!

File: src/cargo/sources/path.rs (hunk 2 above)
Comment (alexcrichton) [inline src/cargo/sources/path.rs]: Ok thanks for checking, sounds good to me!

**Comment by alexcrichton** [comment]
This is looking great to me, feel free to just ping me when CI is passing and I'll r+!

**Comment by behnam** [comment]
We have fixed `ignore` in version `0.2.2`) and all the tests are passing now with expected behavior, which is, for `exclude` field, we will only have additional exclusion, but no patterns changing in the other direction.

Also, I have updated the warning text to inform users with the context of the change in a few words. Here's an example:

```
"Pattern matching for Cargo's include/exclude fields is changing and \
file `{}` WILL be excluded in the next Cargo version.\n\
See https://github.com/rust-lang/cargo/issues/4268 for more info",
```

I think we're good to go, if nothing goes wrong with CI. I will submit a diff for Stage 2—to be landed after the next release. Thanks, @alexcrichton, for all the prompt responses and awesome feedbacks!

**Comment by behnam** [comment]
Uh... because of list of paths not being deterministic, we're getting this kind of failures:

```
failures:
---- exclude stdout ----
	running `/home/travis/build/rust-lang/cargo/target/debug/cargo package --no-verify -v`
thread 'exclude' panicked at '
Expected: execs
    but: differences:
  3 - |[WARNING] [..] file `dir_root_1[/]some_dir[/]file` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `dir_root_3/some_dir/file` WILL be excluded in the next Cargo version.|
  5 - |[WARNING] [..] file `file_root_2` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `dir_root_2/some_dir/file` WILL be excluded in the next Cargo version.|
  7 - |[WARNING] [..] file `dir_root_2[/]some_dir[/]file` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `dir_root_1/some_dir/file` WILL be excluded in the next Cargo version.|
  9 - |[WARNING] [..] file `some_dir[/]dir_deep_1[/]some_dir[/]file` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `some_dir/dir_deep_5/some_dir/file` WILL be excluded in the next Cargo version.|
 13 - |[WARNING] [..] file `some_dir[/]dir_deep_5[/]some_dir[/]file` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `some_dir/dir_deep_1/some_dir/file` WILL be excluded in the next Cargo version.|
 19 - |[WARNING] [..] file `dir_root_3[/]some_dir[/]file` WILL be excluded [..]|
    + |warning: Pattern matching for Cargo's include/exclude fields is changing and file `file_root_2` WILL be excluded in the next Cargo version.|
other output:
``', /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/hamcrest-0.1.1/src/core.rs:31
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    exclude
```

Not sure if it's worth trying to fix this by making the output be deterministic, or just use `[..]` for file names in the test cases. Looking into it...

File: src/cargo/sources/path.rs [hunk 3]
```diff
@@ -311,9 +434,18 @@ impl<'cfg> PathSource<'cfg> {
...
-            let name = dir.file_name().and_then(|s| s.to_str());
+
+        // For package integration tests, we need to sort the paths in a deterministic order to
+        // be able to match stdout warnings in the same order.
+        //
+        // TODO: Drop collect and sort after transition period and dropping wraning tests.
+        // See <https://github.com/rust-lang/cargo/issues/4268>
+        // and <https://github.com/rust-lang/cargo/pull/4270>
+        let mut entries: Vec<fs::DirEntry> = fs::read_dir(path)?.map(|e| e.unwrap()).collect();
+        entries.sort_by(|a, b| a.path().as_os_str().cmp(b.path().as_os_str()));
```
Comment (behnam) [inline src/cargo/sources/path.rs]: To get the warnings be thrown in a deterministic order, I have put this sort method here for the transition period. We can drop it as soon as we drop the checks for warnings.

This will have some extra heap allocation during packaging. However, it does not slow down any build or test functionalities. So, I think it's fine to pay this cost for a few months while making sure correct warnings are thrown for the big change coming.

What do you think, @alexcrichton? Does this sound like a reasonable compromise?

Comment (alexcrichton) [inline src/cargo/sources/path.rs]: Nah yeah seems fine by me!

**Comment by alexcrichton** [comment]
@bors: r+

Thanks @behnam!

**Comment by behnam** [comment]
Tracking issue has the goal and roadmap now: https://github.com/rust-lang/cargo/issues/4268

Next diff is out for adding a notice to the online guide: https://github.com/rust-lang/cargo/pull/4298
